### PR TITLE
diagnostics for RTC resiliency

### PR DESCRIPTION
**Pull request by howard-stearns**
Browser part  of https://docs.google.com/document/d/1je4b-6SycAFkLZcQ0ZOUc0W6gosW6gs529-fn978YlM/edit#heading=h.t2coj8ypvicd

Tasks:
- https://highfidelity.atlassian.net/browse/HIFI-689
- https://highfidelity.atlassian.net/browse/HIFI-690
- https://highfidelity.atlassian.net/browse/HIFI-691

~Not complete yet (see below), but~ There's enough here to facilitate further conversation. Per the doc above, this is reporting TWO log lines per disconnect:
- The lines with labels `ravi...` are fired when RAVI changes state after being connected.
- The line with labels `hifi...` are fired when HiFiCommunicator changes state after being connected, OR if the page visibility changes (on the principle that in the initial version, we are interested in instrumenting session closes from the well-defined case of visible tabs.
The mixer log would be similar, but labeled `mixer...` so that we can easily search and distinguish between all three within a combined multi-source virtual log. I don't know whether we'll need the 'ravi...' version, but it is implemented anyway.

### CODE REVIEW

**Comment by howard-stearns** [comment]
Here are example of three different disconnect cases. I've spread the logs over multiple lines for readability. The initial "columns", up to (but not including) APPSTATE, would also be present in `mixer...` logs.

1. Normal explicit disconnect. Note that RAVISTATE_closed, ONLINE_yes, XPLICITCLOSED_yes, and VISIBLITY_visible

```
2021-06-08T23:40:03.863Z sQf7K56ibe1ipNdepWggZabA5eDDsYpXqCHwfSkF6+Y=
raviRAVISTATE_closed
ravibrowserStatsIP_135.180.172.160 ravibrowserStatsTYPE_srflx ravibrowserStatsPROTOCOL_udp
ravimixerStatsIP_54.184.223.73 ravimixerStatsTYPE_srflx ravimixerStatsPROTOCOL_udp
raviWebSocket:3 raviRTC_closed raviSIGNALING_closed raviICE_closed raviGATHERING_complete
raviAPPSTATE_Connected raviONLINE_yes raviXPLICITCLOSED_yes raviVISIBLITY_visible
raviDEVICE_undefined raviRATING_4g raviDL:10 raviRTT:0
[Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36]
```

```
2021-06-08T23:40:03.864Z sQf7K56ibe1ipNdepWggZabA5eDDsYpXqCHwfSkF6+Y=
hifiRAVISTATE_closed
hifibrowserStatsIP_135.180.172.160 hifibrowserStatsTYPE_srflx hifibrowserStatsPROTOCOL_udp
hifimixerStatsIP_54.184.223.73 hifimixerStatsTYPE_srflx hifimixerStatsPROTOCOL_udp
hifiWebSocket:3 hifiRTC_closed hifiSIGNALING_closed hifiICE_closed hifiGATHERING_complete
hifiAPPSTATE_Disconnected hifiONLINE_yes hifiXPLICITCLOSED_yes hifiVISIBLITY_visible
hifiDEVICE_undefined hifiRATING_4g hifiDL:10 hifiRTT:0
[Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36]
```

2. After turning off wifi while connected. Of course, this doesn't get reported until you connect again from the same browser, in the same or another session.
Note that RAVISTATE_disconnected, ONLINE_no, XPLICITCLOSED_yes (seems like a bug!), and hifiVISIBLITY_visible.
Also note that there is a discrepancy in APPSTATE because of the difference in the HiFiCommunicator state as the communicator and RAVI states fire.

```
2021-06-08T23:47:20.758Z mnPpcV+vQ4cSdI5jk8BAgH4Jo80Xye4hRPDuZmmB6Eg=
raviRAVISTATE_disconnected
ravibrowserStatsIP_135.180.172.160 ravibrowserStatsTYPE_srflx ravibrowserStatsPROTOCOL_udp
ravimixerStatsIP_54.184.223.73 ravimixerStatsTYPE_srflx ravimixerStatsPROTOCOL_udp
raviWebSocket:1 raviRTC_connected raviSIGNALING_stable raviICE_disconnected raviGATHERING_complete
raviAPPSTATE_Connected raviONLINE_no raviXPLICITCLOSED_no raviVISIBLITY_visible
raviDEVICE_undefined raviRATING_4g raviDL:10 raviRTT:0
[Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36]
```

```
2021-06-08T23:47:20.760Z mnPpcV+vQ4cSdI5jk8BAgH4Jo80Xye4hRPDuZmmB6Eg=
hifiRAVISTATE_disconnected
hifibrowserStatsIP_135.180.172.160 hifibrowserStatsTYPE_srflx hifibrowserStatsPROTOCOL_udp
hifimixerStatsIP_54.184.223.73 hifimixerStatsTYPE_srflx hifimixerStatsPROTOCOL_udp
hifiWebSocket:2 hifiRTC_connected hifiSIGNALING_stable hifiICE_disconnected hifiGATHERING_complete
hifiAPPSTATE_Disconnected hifiONLINE_no hifiXPLICITCLOSED_yes hifiVISIBLITY_visible
hifiDEVICE_undefined hifiRATING_4g hifiDL:10 hifiRTT:0
[Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36]
```

3. After closing the tab while connected. This DOES report immediately for hifi (because the definition of the hifi diagnostic is to fire when session ends OR tab becomes invisible), and never fires for ravi diagnostic (not even next session) because the definition there is to fire on ravi state change and ravi never gets the chance.
Note that RAVISTATE_connected, ONLINE_yes, XPLICITCLOSED_no, VISIBLITY_hidden

```
2021-06-08T23:57:35.510Z UsN36mID8+q13kxCRLUa0Vo0McflsSaRmDri0cP4Yxg=
hifiRAVISTATE_connected
hifibrowserStatsIP_135.180.172.160 hifibrowserStatsTYPE_srflx hifibrowserStatsPROTOCOL_udp
hifimixerStatsIP_54.184.223.73 hifimixerStatsTYPE_srflx hifimixerStatsPROTOCOL_udp
hifiWebSocket:1 hifiRTC_connected hifiSIGNALING_stable hifiICE_connected hifiGATHERING_complete
hifiAPPSTATE_Connected hifiONLINE_yes hifiXPLICITCLOSED_no hifiVISIBLITY_hidden
hifiDEVICE_undefined hifiRATING_4g hifiDL:10 hifiRTT:50
[Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36]
```

**Comment by howard-stearns** [comment]
TODO:
- look at what sentry and honeycomb (others?) offer with regard to phoning home.
- meeting Thursday on "Follow up on HIFI-695: create endpoint for client reporting"
- number of users present
- rtc info on dropped packets, etc.

File: src/libravi/RaviSession.ts [hunk 1]
```diff
@@ -472,7 +472,7 @@ export class RaviSession {
   * @internal
  * Constants used as the default filter for the stats collected in the RaviStatsWatcher
  */
-const STATS_WATCHER_FILTER = new Map([
+export const STATS_WATCHER_FILTER = new Map([
```
Comment (howard-stearns) [inline src/libravi/RaviSession.ts:475]: make it available for other code (like mine) to modify.

File: src/classes/HiFiMixerSession.ts [hunk 2]
```diff
@@ -897,6 +914,9 @@ export class HiFiMixerSession {
                 }
                 this._setCurrentHiFiConnectionState(HiFiConnectionStates.Failed);
                 break;
+            case RaviSessionStates.CLOSED:
+               this._setCurrentHiFiConnectionState(HiFiConnectionStates.Disconnected);
+               break;
```
Comment (howard-stearns) [inline src/classes/HiFiMixerSession.ts:919]: This is a functional-change / bug-fix.  Without this applications (and my code) never gets to learn that we closed!

File: src/classes/HiFiMixerSession.ts [hunk 3]
```diff
@@ -583,7 +594,11 @@ export class HiFiMixerSession {
      * Disconnects from the Mixer. Closes the RAVI Signaling Connection and the RAVI Session.
      * @returns A Promise that _always_ Resolves with a "success" status string.
      */
-    async disconnectFromHiFiMixer(): Promise<string> {
+    async disconnectFromHiFiMixer(internal:boolean = false): Promise<string> {
+        if (!internal) {
+            this._raviDiagnostics.noteExplicitApplicationClose();
+            this._hifiDiagnostics.noteExplicitApplicationClose();
+        }
```
Comment (howard-stearns) [inline src/classes/HiFiMixerSession.ts:601]: I'm not crazy about this. I want the report to distinguish between the application calling HiFiCommunicator.disconnectFromAudioAPIServer, vs our own internal calls to here. So I added a an internal:boolean argument and decorated our internal calls. Alternative suggestions welcome.

File: src/classes/HiFiMixerSession.ts [hunk 4]
```diff
@@ -252,7 +261,7 @@ export class HiFiMixerSession {
             let initTimeout = setTimeout(async () => {
                 let errMsg = `Couldn't connect to mixer: Call to \`init\` timed out!`
                 try {
-                    await this.disconnectFromHiFiMixer();
+                    await this.disconnectFromHiFiMixer(true);
```
Comment (howard-stearns) [inline src/classes/HiFiMixerSession.ts:264]: There are several of these. See [L597](https://github.com/highfidelity/hifi-spatial-audio-js/pull/128/files#diff-119c1e67189140964dcdf8544a54d44ab7517ac1354427b54431056b718bf94bR597), below

**Comment by hd23408** [comment]
Is it worth pulling the "diagnostics.ts" bits out into their own repository, do you think? I'm imagining we may want/need to provide our users with the option of opting out of this, and it would be nice to be able to give them a build that explicitly doesn't even include the functionality. But I don't feel super strongly about it -- having it in its own directory like you've got it is probably plenty.

**Comment by howard-stearns** [comment]
I'm imagining/hoping that we'll collect, say, two weeks of data and turn it off. Then based on what we know, we'll make changes and collect another two weeks, or something like that. I don't think it's worth getting too fancy for something like that. If we decide to do something like this permanently, then it would definitely need various options to be controlled by apps.

File: src/classes/HiFiMixerSession.ts (hunk 3 above)
Comment (howard-stearns) [reply src/classes/HiFiMixerSession.ts:601]: yes, I like your suggestion better than mine.... and dislike them both.

File: src/libravi/RaviSession.ts (hunk 1 above)
Comment (howard-stearns) [reply src/libravi/RaviSession.ts:475]: "const" is a misleading name. It has come to mean "the _value_ here, which may be a container, will not change identity in the === sense".  Some languages have an additional term, such as "sealed" or "frozen", to mean "the _content_ here will not change in the == (or mumbleEqual) sense".  There are yet more possibilities, such as "final" to indicate a definition will not be extended (such as by subclassing).  The potential definition space is as big (or bigger?) as the for definitions of equality. (Common Lisp having ==, eq, eql, equal, and equalp.)

File: src/classes/HiFiMixerSession.ts (hunk 2 above)
Comment (hd23408) [reply src/classes/HiFiMixerSession.ts:919]: This change is in tip of main as of https://github.com/highfidelity/hifi-spatial-audio-js/pull/129 if you want to pull that in.

File: src/classes/HiFiMixerSession.ts (hunk 3 above)
Comment (howard-stearns) [reply src/classes/HiFiMixerSession.ts]: done

**Comment by howard-stearns** [comment]
Here are the same three disconnects as above, in the "final"(?) format. Again newlines with a single entry just for clarity here, and would not be used in production.

1. Normal disconnect:

```
021-06-11T22:52:58.135Z 6/0A6TBukdocAmcmxRx68Im1LgSPI7kuXHIO84MtPyw=
 ravibrowserStatsIP_135.180.172.160 ravibrowserStatsTYPE_srflx ravibrowserStatsPROTOCOL_udp
 ravimixerStatsIP_54.202.10.212 ravimixerStatsTYPE_srflx ravimixerStatsPROTOCOL_udp
 ravioutbound-rtp_retransmittedPacketsSent:0 ravioutbound-rtp_packetsSent:50
 raviinbound-rtp_packetsLost:0 raviinbound-rtp_packetsReceived:50 raviinbound-rtp_jitter:0
 raviremote-inbound-rtp_packetsLost:0 raviremote-inbound-rtp_roundTripTime:0.028 raviremote-inbound-rtp_totalRoundTripTime:0.055 raviremote-inbound-rtp_jitter:0.003958333333333334
 raviWebSocket:3 raviRTC_closed raviSIGNALING_closed raviICE_undefined raviGATHERING_complete
 raviCONCURRENCY:1
 raviAPPSTATE_Connected raviRAVISTATE_closed raviONLINE_yes raviXPLICITCLOSED_yes raviVISIBLITY_visible
 raviDEVICE_undefined raviRATING_4g raviDL:5.65 raviRTT:0
 [Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.101 Safari/537.36]
```

```
2021-06-11T22:52:58.136Z 6/0A6TBukdocAmcmxRx68Im1LgSPI7kuXHIO84MtPyw=
 appbrowserStatsIP_135.180.172.160 appbrowserStatsTYPE_srflx appbrowserStatsPROTOCOL_udp
 appmixerStatsIP_54.202.10.212 appmixerStatsTYPE_srflx appmixerStatsPROTOCOL_udp
 appoutbound-rtp_retransmittedPacketsSent:0 appoutbound-rtp_packetsSent:50
 appinbound-rtp_packetsLost:0 appinbound-rtp_packetsReceived:50 appinbound-rtp_jitter:0
 appremote-inbound-rtp_packetsLost:0 appremote-inbound-rtp_roundTripTime:0.028 appremote-inbound-rtp_totalRoundTripTime:0.055 appremote-inbound-rtp_jitter:0.003958333333333334
 appWebSocket:3 appRTC_closed appSIGNALING_closed appICE_undefined appGATHERING_complete
 appCONCURRENCY:1
 appAPPSTATE_Disconnected appRAVISTATE_closed appONLINE_yes appXPLICITCLOSED_yes appVISIBLITY_visible
 appDEVICE_undefined appRATING_4g appDL:5.65 appRTT:0
 [Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.101 Safari/537.36]
```
Turning off wifi:

```
2021-06-11T22:57:21.158Z CZDMvjQDvDvaYcK6Vndq2dHrQNxiPhQeQIIDRv+E4lg=
 ravibrowserStatsIP_135.180.172.160 ravibrowserStatsTYPE_srflx ravibrowserStatsPROTOCOL_udp
 ravimixerStatsIP_54.202.10.212 ravimixerStatsTYPE_srflx ravimixerStatsPROTOCOL_udp
 ravioutbound-rtp_retransmittedPacketsSent:0 ravioutbound-rtp_packetsSent:50
 raviinbound-rtp_packetsLost:0 raviinbound-rtp_packetsReceived:0 raviinbound-rtp_jitter:0.01
 raviremote-inbound-rtp_packetsLost:0 raviremote-inbound-rtp_roundTripTime:0.16 raviremote-inbound-rtp_totalRoundTripTime:0.24 raviremote-inbound-rtp_jitter:0.02266666666666667
 raviWebSocket:1 raviRTC_connected raviSIGNALING_stable raviICE_undefined raviGATHERING_complete
 raviCONCURRENCY:1
 raviAPPSTATE_Connected raviRAVISTATE_disconnected raviONLINE_no raviXPLICITCLOSED_no raviVISIBLITY_visible
 raviDEVICE_undefined raviRATING_4g raviDL:10 raviRTT:0
 [Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.101 Safari/537.36]
```

```
2021-06-11T22:57:21.160Z CZDMvjQDvDvaYcK6Vndq2dHrQNxiPhQeQIIDRv+E4lg=
 appbrowserStatsIP_135.180.172.160 appbrowserStatsTYPE_srflx appbrowserStatsPROTOCOL_udp
 appmixerStatsIP_54.202.10.212 appmixerStatsTYPE_srflx appmixerStatsPROTOCOL_udp
 appoutbound-rtp_retransmittedPacketsSent:0 appoutbound-rtp_packetsSent:50
 appinbound-rtp_packetsLost:0 appinbound-rtp_packetsReceived:0 appinbound-rtp_jitter:0.01
 appremote-inbound-rtp_packetsLost:0 appremote-inbound-rtp_roundTripTime:0.16 appremote-inbound-rtp_totalRoundTripTime:0.24 appremote-inbound-rtp_jitter:0.02266666666666667
 appWebSocket:2 appRTC_connected appSIGNALING_stable appICE_undefined appGATHERING_complete
 appCONCURRENCY:1
 appAPPSTATE_Disconnected appRAVISTATE_disconnected appONLINE_no appXPLICITCLOSED_yes appVISIBLITY_visible
 appDEVICE_undefined appRATING_4g appDL:10 appRTT:0
 [Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.101 Safari/537.36]
```

Closing the tab:

```
2021-06-11T23:01:52.997Z jnSVO8FREfJ7kQfBQLYQy/+2v8I2vk1saD23XFyeMIY=
 appbrowserStatsIP_135.180.172.160 appbrowserStatsTYPE_srflx appbrowserStatsPROTOCOL_udp
 appmixerStatsIP_54.202.10.212 appmixerStatsTYPE_srflx appmixerStatsPROTOCOL_udp
 appoutbound-rtp_retransmittedPacketsSent:0 appoutbound-rtp_packetsSent:54
 appinbound-rtp_packetsLost:0 appinbound-rtp_packetsReceived:50 appinbound-rtp_jitter:0
 appremote-inbound-rtp_packetsLost:0 appremote-inbound-rtp_roundTripTime:0.026 appremote-inbound-rtp_totalRoundTripTime:0.156 appremote-inbound-rtp_jitter:0.00125
 appWebSocket:1 appRTC_connected appSIGNALING_stable appICE_undefined appGATHERING_complete
 appCONCURRENCY:1
 appAPPSTATE_Connected appRAVISTATE_connected appONLINE_yes appXPLICITCLOSED_no appVISIBLITY_hidden
 appDEVICE_undefined appRATING_4g appDL:5.65 appRTT:0
 [Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.101 Safari/537.36]
```

**Comment by howard-stearns** [comment]
Now has real destination url, and works in browser and in node.
But in node, it will not report (and not log anything) if node-fetch is not installed.

Here's an example of running testing bot (for which there is a new PR for a bug fix):
`npm run bots -- --secret FIXME --app_id 11b69b11-1381-4160-af11-55906a55be6e --space_id 85f5144d-05cc-45e9-a69b-b68ecdd1c0a0 --numBots 1 --audio bensound-summer.mp3 --runtimeSeconds 5 --shutdownSeconds 1`

File: src/libravi/RaviSession.ts [hunk 5]
```diff
@@ -621,7 +621,7 @@ class RaviWebRTCStatsWatcher {
     } else {
       // no observers, make sure we are stopped
       if (this._interval) {
-        window.clearInterval(this._interval);
+        clearInterval(this._interval);
```
Comment (howard-stearns) [inline src/libravi/RaviSession.ts:624]: This references to window are both unnecessary, and they break use within node!

File: src/classes/HiFiMixerSession.ts [hunk 6]
```diff
@@ -135,8 +136,8 @@ export class HiFiMixerSession {
      * Since the server doesn't always send the "Provided User ID" in these peer updates, we have to keep track of the (presumably stable) key in `jsonData.peers`
      * associated with that "Provided User ID" in order to forward that "Provided User ID" to the Subscription handler and thus to the Library user.
      * 
-     * And since we are caching that one value, we are also caching the full state for all kwnon peers.
-     * This allows to optimize the received stream of changed data for a given peer from the server to just the necessary bits
+     * And since we are caching that one value, we are also caching the full state for all knwon peers.
```
Comment (hd23408) [inline src/classes/HiFiMixerSession.ts:139]: typo while you're in here :-)

**Review by hd23408 — APPROVED** [review]
This all makes sense to me -- I'm good with merging it once testing is complete. Two high level comments: 

1. Please back-port the RAVI changes to https://github.com/highfidelity/libravi/ as well

2. It might be worth our adding global opt-out flag to the HiFiCommunicator that would allow for opting out of this logging at any point. When we deploy this we will probably want to alert our customers that we're going to be gathering diagnostics in an attempt to improve our connection handling. I could imagine one of them (especially in Europe :-) ) asking how to turn that off.

**Comment by howard-stearns** [comment]
Your're rasing some interesting questions !

Re 1. What is our plan for eventually not maintaining two versions of Ravi JS, and what is the point of the one in Ravi until then?  For example, let's say we eventually want to split it out Ravi JS to its own repo. At that time, we'll presumably want to diff the version in Ravi vs the one here, rather than relying on everyone maintaining everything the same all along. Given that necessary step, why bother backporting changes in the mean time? Doesn't that increase the chances of error (too backport "twice") rather than decreasing it?  What do use the Ravi version of the JS for?  Two counter-proposals:
   a) (Straw-person) Split out the JS portion now, using the version from this repo, and also a Ravi swift client, and a Ravi Unity client, and modify both this repo, Ravi, Swift and Unity to use that new sub-repo(s).
   b) Delete the JS client from the Ravi repo now, and _just_ use the JS client in this repo. (Just as we are already doing with the Swift and Unity API repos, which have their own "ravi" client code included, and which do not duplicate their code within the Ravi repo.

**Review by hd23408 — APPROVED** [review]

**Comment by howard-stearns** [comment]
Closed in favor of https://github.com/highfidelity/hifi-spatial-audio-js/pull/144

**Pull request closed without merging**